### ISSUE TITLE
Data model: Discussion about the initial version of our data model

### ISSUE THREAD

**Issue by tgraupne** [CoronaSupportNetwork/CoronaHelpBroker#1]
During our last call, we were discussing different option and we would like to used this issue as a basis for further discussions and eventually, an initial version.

**Comment by tgraupne** [comment]
Proposal 1:

```json
{
    "icon": "icon-url",
    "platform": "quarantaenehelden",
    "deeplink": "url",
    "postings": [
        {
            "category": "ask-for-help",
            "create-timestamp": 123456789,
            "location": {
                "lat": lat,
                "lng": lng,
                "zip": zip,
                "country": "de"
            },
            "description": "text",
            "notifications": 0,
            "answers": 0,
            "state": "not solved"
        },
        {
            "category": "helper",
            "create-timestamp": 123456789,
            "name": "John Doe",
            "bio": "that's me",
            "contact": {
                "email": "mail@johnd.oe",
                "phone": 1234567890
            },
            "location": {
                "lat": lat,
                "lng": lng,
                "zip": zip,
                "country": "de"
            }
        }
    ]
}
```

**Comment by c-rath** [comment]
Thanks for the good start Tom! 
Do we want to share personal information for the helper category or would it also be sufficient to add a deep link to the referring platform and so avoid GDPR issues?

**Comment by lmoshuber** [comment]
I don't quite understand what the "postings" are. Why are help seekers and helpers combined? And why do help seekers and helpers have different metadata?  It would also be important to define which data are optional and which are not. Or which ones are not possible because of GPDR.
A "standardization" of the tasks might also be useful, but it is not necessary. A string for the description should also serve the purpose, even if it does not allow a more concrete search.

We, all of us who want to be involved in the definition of the interface, should make another call, I think we all still have a little bit different opinion on how the interface should be used in the end.

**Comment by lmoshuber** [comment]
In my opinion every platform should provide two APIs (/helpers & /heelpseekers), which should be crawled from a meta instance in an interval (maybe every hour?) to create the meta database with the combined data. In this way, every platform remains in control of its data and the data is updated regularly (e.g. to update status).

/helpers
```
{
    "icon": "icon-url",
    "platform": "quarantaenehelden",
    "url-to-platform": "https://www.quarantaenehelden.org/"
    "last-update: 012345689, // timestamp
    "helpers": [
        {
            "internal-id": "gsfdsf3434sds",
            "create-timestamp": 123456789,
            "deep-link": "https://www.quarantaenehelden.org/helper/gsfdsf3434sds",
            "location": {
                "lat": lat,
                "lng": lng,
                "zip": zip,
                "country": "de"
            },
            "description": "text",
            "state": "not solved" // should be standardized (ENUM)
        }
    ]
}
```

/helpseekers
```
{
    "icon": "icon-url",
    "platform": "quarantaenehelden",
    "url-to-platform": "https://www.quarantaenehelden.org/"
    "last-update: 012345689, // timestamp
    "helpseekers": [
        {
            "internal-id": "gsfdsf3434sds",
            "create-timestamp": 123456789,
            "deep-link": "https://www.quarantaenehelden.org/helper/gsfdsf3434sds",
            "location": {
                "lat": lat,
                "lng": lng,
                "zip": zip,
                "country": "de"
            },
            "description": "text",
            "state": "not solved" // should be standardized (ENUM)
        }
    ]
}
```

**Comment by tgraupne** [comment]
Hey @lmoshuber, you made some valid points. I added your and my updated proposal to the repository, because I thought it might get easier to discuss this because we could actually work with updated to current proposals and see the diff.

**To answer your questions:**  f9c1dd35c3e85fea24211ed328ba1854b3540e8a
1. I thought we could see every item on our platforms, helpers and help seekers as a "post". Hence, an update of our current posts could be a list of posts.

2. The properties of both categories would be different because the use case is different. A helper does not have e.g. `"notifications": 0`, `"answers": 0` or `"state": "not solved"`.

3. In my most recent version, I added `// optional` to indicate what could or should optional in my opinion. From a GDPR perspective, every platform seems to have a different approach and your are right, every platform should be in control of their data and decides what to share.

**Now I'd like to say something about your proposal:** f9c1dd35c3e85fea24211ed328ba1854b3540e8a
1. I am totally fine with separating helpers and help seekers.

2. I would be also ok with your proposed _Pull-Approach_, where every platform provides an API and the central system pulls information.

2.1. However, I would also propose a _Push-Approach_ as an alternative, where platforms push there most up-to-date data to the central system.

The **benefits** of that would be, that platforms would **not** need to implement an API and we could distribute updates fast.
One **disadvantage** would be, that the Pull-Approach enables the central system to consume data in it's own speed.
And ss a **neutral** point, every platform would be in charge of their data in both scenarios.

I think both scenarios have their advantages and disadvantages and at the end of the day it is just one of many decisions we have to make. 🤷‍♀️

However, in a _Push-Approach-Scenario_, once we settled on a data model, our "centralised system" could be a simple but scalable document-store, e.g. etcd.

**Comment by lmoshuber** [comment]
You are right, both options have both advantages and disadvantages. But we're already talking about updating the whole dataset at once, not just individual datasets, right?

As for the properties - how they are composed depends on the process of the respective platform ... Here, all the platforms involved should say something about this.

**Comment by tgraupne** [comment]
> You are right, both options have both advantages and disadvantages. But we're already talking about updating the whole dataset at once, not just individual datasets, right?

🤔 in my humble opinion, not necessarily. I think we should provide the option to update single records as well as multiple records in one "batch". But neither should be a problem with both approaches, _Push_ and _Pull_.

> As for the properties - how they are composed depends on the process of the respective platform ... Here, all the platforms involved should say something about this.

That is true, I just asked for further feedback in slack. 🤞
